### PR TITLE
Fix quantization

### PR DESCRIPTION
**Pull request by b-flo**
This PR fixes dynamic quantization for Transducer (and also update arguments doc). There are currently two issues that ends-up with an `AttributeError` or `RuntimeError` : 

1) With `quantize-config: [torch.nn.LSTM]` and any LSTM-based encoder
2) With `quantize-config: [torch.nn.Linear]` and most search algorithms (greedy, default, NSC, and mAES)

The first issue is due to `flatten_parameters` attribute. Looking at the [torch doc](https://pytorch.org/docs/stable/_modules/torch/nn/quantized/dynamic/modules/rnn.html#LSTM), it seems `torch.nn.quantized.dynamic.LSTM` doesn't have the attribute compared to `torch.nn.quantized.LSTM`. I suppose other models' encoders are impacted by this?
The second issue is due to passing tensors with dim() < 2 to `torch.nn.quantized.dynamic.Linear`. Typically, when we rely on non-batch beam (greedy or default) or during prefix search (default, NSC and mAES).

The fixes are straightforward, It's a bit ugly. Feel free to propose something else!
Also, I guess I added too many tests, some may be useless. I'll check that out.

### CODE REVIEW

**Comment by b-flo** [comment]
Hum, it seems there are additional issues : 

- With `quantize-config: [torch.nn.LSTM]` and torch < 1.4
- With `quantize-dtype: float16` and torch < 1.5

The first issue is related to using a packed sequence with quantized LSTM (see code notes). For the second one, I'm not sure, I couldn't find anything useful as a reference.

From what I gather, all mentioned issues are not checked in the integration tests or unit tests. I suppose, outside the one related to Transducer `JointNetwork`, others ASR models should be impacted by these issues.

@sw005320 @ShigekiKarita Related topic: should we plan to bump the torch requirements for ESPnet1's next major version (if there is one)? I'm afraid it will become difficult to add some features and maintain all torch versions (1.3.1 - 1.9+) without overloading/impacting some parts of the current implementation.

**Comment by ShigekiKarita** [comment]
> With quantize-dtype: float16 and torch < 1.5

> @sw005320 @ShigekiKarita Related topic: should we plan to bump the torch requirements for ESPnet1's next major version

Sounds good. Pytorch 1.5 was released [one year ago](https://pytorch.org/blog/pytorch-1-dot-5-released-with-new-and-updated-apis/). I believe many people already use it.

File: espnet/asr/pytorch_backend/asr.py [hunk 1]
```diff
@@ -986,12 +987,32 @@ def recog(args):
...
+            and "lstm" in train_args.etype
+            and torch.nn.LSTM in q_config
+        ):
+            raise ValueError(
+                "Quantized LSTM in ESPnet is only supported with torch 1.4+."
+            )
+
+        # Dunno why but weight_observer from dynamic quantized module must have
+        # dtype=torch.qint8 with torch < 1.5 although dtype=torch.float16 is supported.
+        if args.quantize_dtype == "float16" and LooseVersion(
```
Comment (ShigekiKarita) [inline espnet/asr/pytorch_backend/asr.py:1005]: Can you test this new option in https://github.com/espnet/espnet/blob/master/ci/test_integration_espnet1.sh?

File: espnet/asr/pytorch_backend/asr.py [hunk 2]
```diff
@@ -986,12 +987,32 @@ def recog(args):
...
+            torch.__version__
+        ) < LooseVersion("1.5.0"):
+            logging.warning(
```
Comment (ShigekiKarita) [inline espnet/asr/pytorch_backend/asr.py:1008]: Can you make this `ValueError` instead of warning because it is critical when someone compares performance? Of course, if you update the requirement to 1.5, you should remove this block.

**Review by ShigekiKarita — APPROVED** [review]

File: espnet/asr/pytorch_backend/asr.py (hunk 1 above)
Comment (b-flo) [reply espnet/asr/pytorch_backend/asr.py:1005]: Sorry, I'm a bit confused. If I add an integration test for `quantize-dtype: float16`, CI will fail because `ValueError `will be raised?
Ultimately, I would need to add a unit test but we don't directly test `recog(...)` from `asr.py`...

Comment (klknn) [reply espnet/asr/pytorch_backend/asr.py:1005]: I see. So let's go pytorch 1.5+ for the integration test? @sw005320 

I agree that the unittest is very useful, and be careful with pytest.timeout.

**Comment by klknn** [comment]
Because we implemented several new features in --api=v2, can you also update https://github.com/espnet/espnet/blob/master/espnet/asr/pytorch_backend/recog.py

**Comment by ShigekiKarita** [comment]
(Sorry I used my different account in the previous reply)

Can you update https://github.com/espnet/espnet/blob/master/espnet/asr/pytorch_backend/recog.py#L22

File: espnet/asr/pytorch_backend/asr.py (hunk 1 above)
Comment (ShigekiKarita) [reply espnet/asr/pytorch_backend/asr.py:1005]: > CI will fail because ValueError will be raised?

Shall we omit pytorch 1.5- in the CI? @sw005320 

> unittest

Sounds good, and be careful with pytest.timeout.

**Comment by b-flo** [comment]
> Can you also update recog_v2 https://github.com/espnet/espnet/blob/master/espnet/asr/pytorch_backend/recog.py#L22?

Sure, I'll do!
Btw, what about the first issue (with `quantize-config: [torch.nn.LSTM]` and any LSTM-based encoder)? If I'm not mistaken other models are impacted and I didn't fix it RNN/RNNP encoders outside Transducer model. I'm not too fan of skipping flatten_parameters() in these cases but I'm also not sure we can fix it differently.

File: espnet/asr/pytorch_backend/asr.py (hunk 1 above)
Comment (b-flo) [reply espnet/asr/pytorch_backend/asr.py:1005]: > Sounds good, and be careful with pytest.timeout

Oh btw, how can I be sure which value I should assign to `pytest.mark.execution_timeout(...)`? I mean, I'm setting it up according to my local settings but CI rely on different hardware.

File: test/test_e2e_asr_transducer.py [hunk 3]
```diff
@@ -391,3 +395,86 @@ def test_invalid_aux_transducer_loss_enc_layers():
...
+    ):
+        # In recognize(...) from asr.py we raise ValueError however
+        # AssertionError is originaly raised by torch.
+        with pytest.raises(AssertionError):
+            model = torch.quantization.quantize_dynamic(
+                model,
+                quantize_dic["mod"],
+                dtype=quantize_dic["dtype"],
+            )
+        pytest.skip("Skip rest of the test after checking AssertionError")
```
Comment (b-flo) [inline test/test_e2e_asr_transducer.py:449]: A `pytest.skip` is needed here otherwise next `pytest.raises(...)` may fail because `quantize_dynamic` was not applied.
Note that I'm fully aware this is bad practice to do that however this is only a temporary measure, it will be updated after we bump torch version.

**Comment by sw005320** [comment]
> > With quantize-dtype: float16 and torch < 1.5
> 
> > @sw005320 @ShigekiKarita Related topic: should we plan to bump the torch requirements for ESPnet1's next major version
> 
> Sounds good. Pytorch 1.5 was released [one year ago](https://pytorch.org/blog/pytorch-1-dot-5-released-with-new-and-updated-apis/). I believe many people already use it.

I agree.
Make Pytorch >=1.5 as default should be no problem.

File: espnet/asr/pytorch_backend/asr.py (hunk 1 above)
Comment (ShigekiKarita) [reply espnet/asr/pytorch_backend/asr.py:1005]: Because it is a kinda integration test, you can set a large value enough to pass CI. For example, 50 sec is safe:
https://github.com/espnet/espnet/blob/08feae5bb93fa8f6dcba66760c8617a4b5e39d70/test/espnet2/tasks/test_abs_task.py#L108-L109

**Comment by b-flo** [comment]
OK, I added the checks to API v2. I guess we can keep it until we bump torch version. However : 

- I didn't test nor add checks for quantization in others implementations (LM, TTS, ST, etc). Depending on the PR for torch version, I'll add them in the meantime.
- The first issue (with quantize-config: [torch.nn.LSTM] and any LSTM-based encoder) happen even with to torch 1.9.  We should maybe think about some other fix/implementation: 1) I'm not fan of overloading some computation parts with if-else, and 2) we could remove calls to `flatten_parameters()` and it would also fix some issues we have with multi-GPUs (cc @Desh in case the conversation go somewhere). Obviously, it will impact training time.

**Comment by sw005320** [comment]
Thanks, @b-flo!
